### DISCUSSION TITLE
Transcript of knowledge sharing session on modern dice

### DISCUSSION THREAD

**Discussion by Nero5023** [facebook/buck2#807] [General]
We put the transcript of our knowledge sharing session on modern DICE here https://buck2.build/docs/insights_and_knowledge/modern_dice/

Welcome to read, comment and ask questions. 

(DICE is our buck2 underlying computation engine to manage complex incremental computations.)

**Comment by cbarrete** [comment]
Hey, thanks for the update! It's really useful for the outside world to learn about what's going on internally, especially in the absence of release notes :)
Quick question: this presentation mentions early cutoff. It was my understanding ~a year ago that it wasn't implemented. Is it available now?

**Reply by Nero5023** [reply]
Yes, in dice we have early cutoff. I am not familiar with the history and I don't know if we point to the same thing.
Let me cc @cjhopman

**Reply by cjhopman** [reply]
Yeah DICE has early cutoff. It depends on the user implementing Key::Value equality. For buck, there's a lot of different scenarios that can hit dice early cutoff, artifacts have equality, so consumers of those can potentially get early cutoff, buckconfig values also have it, so we see it there.

One notable thing that doesn't have equality implemented would be analysis results, including action values. This means that a change that causes analysis to re-run will cause us to not get any early cutoff in the produced actions even if they logically didn't change. In that scenario, there are other (non-DICE) mechanisms that can get the equivalent benefit (e.g. depfile-based action matches).

**Reply by cbarrete** [reply]
Thanks, that's useful information.

Does the prelude support depfiles, or is it something that every project needs to implement for itself? https://buck2.build/docs/rule_authors/dep_files/ is quite generic and it's not clear to me e.g. how to benefit from them in an existing C++ build.
